### PR TITLE
[DO NOT MERGE] Fix CMake static build

### LINKED ISSUE #110238: PyTorch with non-shared build (building a single shared lib) is unsupported
### 🐛 Describe the bug

Building with `BUILD_SHARED_LIBS=0` to produce a single `libtorch.so` appears to be broken when using Python/torch rather than C++/libtorch.

Full build command:
```
BUILD_SHARED_LIBS=0 TORCH_CUDA_ARCH_LIST="9.0" CUDA_HOME="/usr/local/cuda"     CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" USE_SYSTEM_NCCL=1 USE_MAGMA=1 USE_OPENCV=1 BUILD_CAFFE2=0 BUILD_CAFFE2_OPS=0 BUILD_TEST=0 USE_CUSPARSELT=1 USE_FBGEMM=0 USE_KINETO=0 USE_METAL=0 USE_NATIVE_ARCH=1 USE_NNPACK=0 USE_QNNPACK=0 USE_VALGRIND=0 USE_XNNPACK=0 USE_ITT=0 USE_C10D_UCC=0 USE_GLOO=0 USE_C10D_MPI=0 ONNX_ML=0 USE_TENSORPIPE=0 BUILD_NVFUSER=0 BUILD_FUNCTORCH=1 BUILD_SPLIT_CUDA=0 python3 setup.py develop
```

The following output is observed:
```
>>> import torch
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/workspace/pytorch/torch/__init__.py", line 235, in <module>
    _load_global_deps()
  File "/workspace/pytorch/torch/__init__.py", line 194, in _load_global_deps
    raise err
  File "/workspace/pytorch/torch/__init__.py", line 175, in _load_global_deps
    ctypes.CDLL(lib_path, mode=ctypes.RTLD_GLOBAL)
  File "/usr/lib/python3.10/ctypes/__init__.py", line 374, in __init__
    self._handle = _dlopen(self._name, mode)
OSError: /workspace/pytorch/torch/lib/libtorch_global_deps.so: cannot open shared object file: No such file or directory
```
Running with ` TORCH_USE_RTLD_GLOBAL=1` to bypass the above failure seems to indicate that something is being loaded twice:
```
>>> import torch
terminate called after throwing an instance of 'c10::Error'
  what():  Tried to register multiple backend fallbacks for the same dispatch key Conjugate; previous registration registered at /workspace/pytorch/aten/src/ATen/ConjugateFallback.cpp:17, new registration registered at /workspace/pytorch/aten/src/ATen/ConjugateFallback.cpp:17
Exception raised from registerFallback at /workspace/pytorch/aten/src/ATen/core/dispatch/Dispatcher.cpp:401 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x98 (0xfffe1caf24d8 in /workspace/pytorch/torch/lib/libshm.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf8 (0xfffe16cbdd28 in /workspace/pytorch/torch/lib/libshm.so)
frame #2: c10::Dispatcher::registerFallback(c10::DispatchKey, c10::KernelFunction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x558 (0xfffe27e16a88 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #3: torch::Library::_fallback(torch::CppFunction&&) & + 0x1c4 (0xfffe27e53e54 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #4: <unknown function> + 0x21b8790 (0xfffe27bd8790 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #5: <unknown function> + 0x21bc00c (0xfffe27bdc00c in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #6: <unknown function> + 0x16824d4 (0xfffe270a24d4 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #7: <unknown function> + 0x5624 (0xfffe37405624 in /lib/ld-linux-aarch64.so.1)
frame #8: <unknown function> + 0x572c (0xfffe3740572c in /lib/ld-linux-aarch64.so.1)
frame #9: _dl_catch_exception + 0xd0 (0xfffe3722d1f0 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #10: <unknown function> + 0xbf5c (0xfffe3740bf5c in /lib/ld-linux-aarch64.so.1)
frame #11: _dl_catch_exception + 0x78 (0xfffe3722d198 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #12: <unknown function> + 0xc2fc (0xfffe3740c2fc in /lib/ld-linux-aarch64.so.1)
frame #13: <unknown function> + 0x796e4 (0xfffe371796e4 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #14: _dl_catch_exception + 0x78 (0xfffe3722d198 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #15: _dl_catch_error + 0x40 (0xfffe3722d260 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #16: <unknown function> + 0x791c0 (0xfffe371791c0 in /usr/lib/aarch64-linux-gnu/libc.so.6)
frame #17: dlopen + 0x54 (0xfffe37179784 in /usr/lib/aarch64-linux-gnu/libc.so.6)
<omitting python frames>

Aborted (core dumped)
```

### Versions

(upstream source build)
```
Collecting environment information...
PyTorch version: N/A
Is debug build: N/A
CUDA used to build PyTorch: N/A
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (aarch64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.27.5
Libc version: glibc-2.35

Python version: 3.10.12 (main, Jun 11 2023, 05:26:28) [GCC 11.4.0] (64-bit runtime)
Python platform: Linux-6.2.0-1008-nvidia-64k-aarch64-with-glibc2.35
Is CUDA available: N/A
CUDA runtime version: 12.2.140
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: GPU 0: NVIDIA Graphics Device
Nvidia driver version: 535.115
```

### PR DESCRIPTION
**Pull request by cyyever**
This PR tries to solve the static build issues in CMake build system. Fixes #110238

### CODE REVIEW

**Review by albanD — APPROVED** [review]
Sounds good to me if all build are green!

**Comment by cyyever** [comment]
@albanD The ROCM failure may be unrelated

**Comment by albanD** [comment]
Since this is a bit risky, I'll import it and make sure it doesn't break build on our end before merging it.

**Comment by albanD** [comment]
Internal build is good. Once CI is green, we can merge here!

**Comment by cyyever** [comment]
@albanD I have more fixes. By the way, if static building is important, should we enable the CI test just to ensure that further changes wouldn't break it again?

File: aten/src/ATen/test/cuda_distributions_test.cu [hunk 1]
```diff
@@ -119,7 +119,6 @@ TEST(DistributionsTest, TestPhiloxIncrementBigUniformTensor) {
   // calculate maximum number of threads that can be launched
   // and set the numel to be 8 times that
   const int block_size = 256;
-  dim3 dim_block(block_size);
```
Comment (albanD) [inline aten/src/ATen/test/cuda_distributions_test.cu:122]: This is just a general cleanup or important?

File: cmake/Dependencies.cmake [hunk 2]
```diff
@@ -1354,23 +1354,24 @@ endif()
 
 if(USE_DISTRIBUTED AND USE_TENSORPIPE)
   if(MSVC)
+    set(USE_TENSORPIPE OFF CACHE BOOL "" FORCE)
```
Comment (albanD) [inline cmake/Dependencies.cmake:1357]: Why not use caffe2_update_option here?

File: cmake/Dependencies.cmake [hunk 3]
```diff
@@ -1354,23 +1354,24 @@ endif()
...
     message(WARNING "Tensorpipe cannot be used on Windows.")
   else()
     if(USE_CUDA)
       set(TP_USE_CUDA ON CACHE BOOL "" FORCE)
       set(TP_ENABLE_CUDA_IPC ON CACHE BOOL "" FORCE)
     endif()
     set(TP_BUILD_LIBUV ON CACHE BOOL "" FORCE)
-    add_compile_options(-DTORCH_USE_LIBUV)
```
Comment (albanD) [inline cmake/Dependencies.cmake:1364]: Did you remove this one because it is already in tensorpipe_uv and this was a duplicate?

**Review by albanD — COMMENTED** [review]
Amazing cleanup job! Thanks for taking the time to do that. Only a couple of clarifying questions.

**Review by malfet — COMMENTED** [review]
Please add a test plan of some sort (as I don't think we can rely on CI for this one) Or better, would it be possible to add test to the CI (because as https://github.com/pytorch/pytorch/issues/110238 suggests this is not tested at all at the moment, is it?)

**Review by malfet — CHANGES_REQUESTED** [review]
Please add a test plan and move cleanup changes to a separate PR (I wanted to submit one removing GCC_7 checks, but I don't think they belong to PR with title `Fix static build`)

**Comment by albanD** [comment]
Doing another import since this changed significantly

File: cmake/Dependencies.cmake (hunk 3 above)
Comment (cyyever) [reply cmake/Dependencies.cmake:1364]: Yes

**Comment by cyyever** [comment]
> Please add a test plan and move cleanup changes to a separate PR (I wanted to submit one removing GCC_7 checks, but I don't think they belong to PR with title `Fix static build`)

Yes, I plan to do more significant changes in this PR because the static build still has problem. As a result, some changes were extracted to #110378

**Comment by cyyever** [comment]
The main obstacle I found is that CUDA and other backends use c10::registery to wrap CUDA and other backend APIs. However, in static build, the code is optimized out in linking so that the resulting so file has only CPU operations. To avoid the complex static linking order issues and cyclic dependency, a better solution is to build torch_cpu and torch_cuda as CMake OBJECT libraries when BUILD_SHARE_LIBS is off. Hopefully, this behavior change wouldn't affect the main cases of building so files and break the code. But we need to re-enable or setup new CI tests for the static building.

File: aten/src/ATen/test/cuda_distributions_test.cu (hunk 1 above)
Comment (cyyever) [reply aten/src/ATen/test/cuda_distributions_test.cu:122]: A general cleanup, should be moved to another PR.

File: cmake/Dependencies.cmake (hunk 2 above)
Comment (cyyever) [reply cmake/Dependencies.cmake:1357]: I forgot, will change it.

**Comment by eqy** [comment]
An update on my end, not sure this is expected but building with the current branch and minimal environment variables
`BUILD_SHARED_LIBS=0 TORCH_CUDA_ARCH_LIST=9.0 CUDA_HOME="/usr/local/cuda"     CMAKE_PREFIX_PATH="$(dirname $(which conda))/../" python3 setup.py develop` 
I see errors relating to undefined references to TensorPipe  e.g.,
```
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0xc6b1): undefined reference to `tensorpipe::Context::listen(std::vector<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::allocator<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > > > const&)'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0xc72f): undefined reference to `tensorpipe::Listener::url(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0xc9c8): undefined reference to `tensorpipe::Listener::accept(std::function<void (tensorpipe::Error const&, std::shared_ptr<tensorpipe::Pipe>)>)'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::TensorPipeAgent(c10::intrusive_ptr<c10d::Store, c10::detail::intrusive_target_default_null_type<c10d::Store> > const&, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, short, c10::optional<int>, torch::distributed::rpc::TensorPipeRpcBackendOptions, std::unordered_map<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > >, std::hash<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::equal_to<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > >, std::allocator<std::pair<std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > > > > >, std::vector<c10::Device, std::allocator<c10::Device> >, std::unique_ptr<torch::distributed::rpc::RequestCallback, std::default_delete<torch::distributed::rpc::RequestCallback> >)':
tensorpipe_agent.cpp:(.text+0xeafa): undefined reference to `tensorpipe::Context::Context(tensorpipe::ContextOptions)'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::respond(std::shared_ptr<tensorpipe::Pipe>&)::{lambda(tensorpipe::Error const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Stream, std::allocator<c10::Stream> >)#1}::operator()(tensorpipe::Error const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Stream, std::allocator<c10::Stream> >)':
tensorpipe_agent.cpp:(.text+0xfb12): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0xfb43): undefined reference to `tensorpipe::Error::what[abi:cxx11]() const'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0xfc13): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::handleClientError(torch::distributed::rpc::TensorPipeAgent::ClientPipe&, tensorpipe::Error const&)':tensorpipe_agent.cpp:(.text+0x1054a): undefined reference to `tensorpipe::Error::what[abi:cxx11]() const'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::send(torch::distributed::rpc::WorkerInfo const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, float, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > > const&)::{lambda(tensorpipe::Error const&)#2}::operator()(tensorpipe::Error const&)':
tensorpipe_agent.cpp:(.text+0x10785): undefined reference to `typeinfo for tensorpipe::PipeClosedError'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10854): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10888): undefined reference to `tensorpipe::Error::what[abi:cxx11]() const'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10967): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::send(torch::distributed::rpc::WorkerInfo const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, float, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > > const&)::{lambda(tensorpipe::Error const&)#2}::operator()(tensorpipe::Error const&)::{lambda(tensorpipe::Error const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Stream, std::allocator<c10::Stream> >)#1}::operator()(tensorpipe::Error const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, std::vector<c10::Stream, std::allocator<c10::Stream> >) const':
tensorpipe_agent.cpp:(.text+0x10a57): undefined reference to `typeinfo for tensorpipe::PipeClosedError'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10af8): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10b29): undefined reference to `tensorpipe::Error::what[abi:cxx11]() const'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x10c1a): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `torch::distributed::rpc::TensorPipeAgent::send(torch::distributed::rpc::WorkerInfo const&, c10::intrusive_ptr<torch::distributed::rpc::Message, c10::detail::intrusive_target_default_null_type<torch::distributed::rpc::Message> >, float, std::unordered_map<c10::Device, c10::Device, std::hash<c10::Device>, std::equal_to<c10::Device>, std::allocator<std::pair<c10::Device const, c10::Device> > > const&)':
tensorpipe_agent.cpp:(.text+0x11b68): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x11ecd): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x12769): undefined reference to `tensorpipe::Pipe::getRemoteName[abi:cxx11]()'
/usr/bin/ld: tensorpipe_agent.cpp:(.text+0x129ab): undefined reference to `tensorpipe::Context::connect(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&, tensorpipe::PipeOptions)'
/usr/bin/ld: /workspace/pytorch/build/lib/libtorch_cpu.a(tensorpipe_agent.cpp.o): in function `std::_Sp_counted_ptr_inplace<tensorpipe::Context, std::allocator<tensorpipe::Context>, (__gnu_cxx::_Lock_policy)2>::_M_dispose()':
tensorpipe_agent.cpp:(.text._ZNSt23_Sp_counted_ptr_inplaceIN10tensorpipe7ContextESaIS1_ELN9__gnu_cxx12_Lock_policyE2EE10_M_disposeEv[_ZNSt23_Sp_counted_ptr_inplaceIN10tensorpipe7ContextESaIS1_ELN9__gnu_cxx12_Lock_policyE2EE10_M_disposeEv]+0x9): undefined reference to `tensorpipe::Context::~Context()'
collect2: error: ld returned 1 exit status
```

**Comment by eqy** [comment]
@cyyever thanks, I'll update what I observe in the PR thread

**Comment by eqy** [comment]
This is what I currently see:
```
>>> import torch
terminate called after throwing an instance of 'c10::Error'
  what():  Tried to register multiple backend fallbacks for the same dispatch key Conjugate; previous registration registered at /workspace/pytorch/aten/src/ATen/ConjugateFallback.cpp:17, new registration registered at /workspace/pytorch/aten/src/ATen/ConjugateFallback.cpp:17
Exception raised from registerFallback at /workspace/pytorch/aten/src/ATen/core/dispatch/Dispatcher.cpp:401 (most recent call first):
frame #0: c10::Error::Error(c10::SourceLocation, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0xae (0x7ff0ae668e1e in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #1: c10::detail::torchCheckFail(char const*, char const*, unsigned int, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) + 0xf3 (0x7ff0a640580b in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #2: c10::Dispatcher::registerFallback(c10::DispatchKey, c10::KernelFunction, std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> >) + 0x72e (0x7ff0a6a88d0e in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #3: torch::Library::_fallback(torch::CppFunction&&) & + 0x25f (0x7ff0a6acde7f in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #4: <unknown function> + 0x161fd7f (0x7ff0a67e9d7f in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #5: <unknown function> + 0x1624bf7 (0x7ff0a67eebf7 in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #6: <unknown function> + 0x124024c (0x7ff0a640a24c in /workspace/pytorch/torch/lib/libnvfuser_codegen.so)
frame #7: <unknown function> + 0x647e (0x7ff12524647e in /lib64/ld-linux-x86-64.so.2)
frame #8: <unknown function> + 0x6568 (0x7ff125246568 in /lib64/ld-linux-x86-64.so.2)
frame #9: _dl_catch_exception + 0xe5 (0x7ff12504cc85 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #10: <unknown function> + 0xdff6 (0x7ff12524dff6 in /lib64/ld-linux-x86-64.so.2)
frame #11: _dl_catch_exception + 0x88 (0x7ff12504cc28 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #12: <unknown function> + 0xe34e (0x7ff12524e34e in /lib64/ld-linux-x86-64.so.2)
frame #13: <unknown function> + 0x906bc (0x7ff124f686bc in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #14: _dl_catch_exception + 0x88 (0x7ff12504cc28 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #15: _dl_catch_error + 0x33 (0x7ff12504ccf3 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #16: <unknown function> + 0x901ae (0x7ff124f681ae in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #17: dlopen + 0x48 (0x7ff124f68748 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #18: at::DynamicLibrary::DynamicLibrary(char const*, char const*, bool) + 0x48 (0x7ff10efbb318 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #19: <unknown function> + 0x674a68c (0x7ff112c4368c in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #20: <unknown function> + 0x2035197 (0x7ff10e52e197 in /workspace/pytorch/torch/lib/libtorch_python.so)
frame #21: <unknown function> + 0x647e (0x7ff12524647e in /lib64/ld-linux-x86-64.so.2)
frame #22: <unknown function> + 0x6568 (0x7ff125246568 in /lib64/ld-linux-x86-64.so.2)
frame #23: _dl_catch_exception + 0xe5 (0x7ff12504cc85 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #24: <unknown function> + 0xdff6 (0x7ff12524dff6 in /lib64/ld-linux-x86-64.so.2)
frame #25: _dl_catch_exception + 0x88 (0x7ff12504cc28 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #26: <unknown function> + 0xe34e (0x7ff12524e34e in /lib64/ld-linux-x86-64.so.2)
frame #27: <unknown function> + 0x906bc (0x7ff124f686bc in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #28: _dl_catch_exception + 0x88 (0x7ff12504cc28 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #29: _dl_catch_error + 0x33 (0x7ff12504ccf3 in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #30: <unknown function> + 0x901ae (0x7ff124f681ae in /usr/lib/x86_64-linux-gnu/libc.so.6)
frame #31: dlopen + 0x48 (0x7ff124f68748 in /usr/lib/x86_64-linux-gnu/libc.so.6)
<omitting python frames>
```

**Comment by cyyever** [comment]
@eqy This is what I described before, some static libs are linked multiple times in static build. Hopefully, we can gradually identify and fix them without the risk of breaking the shared build. Those fixes are being proposed in other PRs.

**Comment by CorbinFoucart** [comment]
> @eqy This is what I described before, some static libs are linked multiple times in static build. Hopefully, we can gradually identify and fix them without the risk of breaking the shared build. Those fixes are being proposed in other PRs.

@cyyever I'm looking to fix this error myself for a static pytorch build with CPU and CUDA support -- how did you go about identifying which static libs are double linked? I don't need to preserve the dynamic lib behavior as this will be a patch.

**Comment by cyyever** [comment]
@CorbinFoucart It was long ago that I forget most details now. But for problematic symbols in static linking you can switch to dynamic linking and use `ldd -r` to observer how they are linked recursively into different libraries. Also I used both gcc and clang to enable a larger set of diagnostic linking warnings.

**Comment by CorbinFoucart** [comment]
@cyyever Thank you, that's ultimately what I ended up doing.

For anyone interested in this approach, note that your final executable will need to link to some of the static torch libraries using `-Wl,--whole-archive`.

**Pull request closed without merging**